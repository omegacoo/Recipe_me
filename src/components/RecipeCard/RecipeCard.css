mark {
    background-color: #F0EAD6;
}

.RecipeCard {
    width: 95%;
    margin-left: auto;
    margin-right: auto;
    margin-bottom: 20px;
    list-style: none;
    text-align: center;
    overflow: hidden;
    border-right: 1px solid gray;
    border-bottom: 1px solid gray;
    border-radius: 15px;
}

.RecipeCard_title {
    border-bottom: 1px solid black;
    font-weight: bolder;
    width: 80%;
    margin-left: auto;
    margin-right: auto;
    font-family: 'Indie Flower', cursive;
    font-size: 3.5vh;
}

.RecipeCard_description  {
    font-size: 2vh;
    margin: 1vh;
}

.RecipeCard_full {
    position: absolute;
    height: 75vh;
    width: 90vw;
    left: 0;
    top: 10vh;
    z-index: 1;
    background-image: url('../../assets/notecard.png');
    background-repeat: no-repeat;
    background-size: 100% 100%;
    margin: 3vh;
    padding: 2vh;
    padding-right: 3vh;
    padding-bottom: 3vh;
    display: grid;
    grid-template-rows: min-content 10% 5% 20% 6%;
}

.title_container {
    font-weight: bolder;
}

.RecipeCard_full_title {
    font-family: 'Indie Flower', cursive;
    font-size: 4vh;
    width: 100%;
    margin-left: auto;
    margin-right: auto;
}

.RecipeCard_full .closePopup {
    background-color: red;
    color: #F0EAD6;
    border-radius: 50%;
    font-size: 3vh;
    position: absolute;
    left: 3vw;
    top: 2vh;
    box-shadow: -3px 0px 3px black;
}

.Recipe_times {
    display: flex;
    flex-direction: row;
    justify-content: center;
    border-top: 1px solid black;
    width: fit-content;
    height: fit-content;
    margin-left: auto;
    margin-right: auto;
}

.prep_time, .cook_time {
    border: 1px solid black;
    border-top: none;
    border-bottom-left-radius: 15px;
    border-bottom-right-radius: 15px;
    padding: .75vh;
}

.prep_time {
    background-color: #f1c40f;
}

.cook_time {
    background-color: #16a085;
}

.Recipe_ingredients_label {
    border-bottom: 2px solid black;
}

.Recipe_ingredients {
    max-height: 95%;
    width: 90%;
    margin-top: 1vh;
    margin-left: auto;
    margin-right: auto;
    list-style: none;
    border: 1px solid black;
    padding: 1vh;
    background-color: #F0EAD6;
    text-align: left;
    overflow-y: auto;
}

.Recipe_ingredients::-webkit-scrollbar,
.Recipe_instructions::-webkit-scrollbar {
    -webkit-appearance: none;
}

.Recipe_ingredients::-webkit-scrollbar:vertical {
    width: 5px;
}

.Recipe_ingredients::-webkit-scrollbar:horizontal {
    width: 11px;
}

.Recipe_ingredients::-webkit-scrollbar-thumb {
    background-color: black;
}

.Recipe_ingredients li {
    display: inline-block;
    width: 47%;
    margin: .2vh;
    margin-right: 1vw;
    border-bottom: .5px solid black;
}

.Recipe_instructions_label {
    border-bottom: 2px solid black;
    margin-top: 1vh;
}

.Recipe_instructions {
    white-space: pre-wrap;
    text-align: left;
    margin: 2vh;
    overflow: auto;
}

/* iPad 768x1024 Portrait */
@media only screen and (min-width: 768px) and (min-height: 1024px) {
    .RecipeCard_full .closePopup {
        width: 4vw;
    }

    .RecipeCard_full_title {
        width: 80%;
    }
}

/* Desktop */
@media screen and (min-width: 1025px) {
    .RecipeCard:hover,
    .RecipeCard_title:hover,
    .RecipeCard_description:hover {
        cursor: pointer;
    }

    .RecipeCard_full {
        width: 50vw;
        left: 25vw;
    }

    .Recipe_instructions {
        margin-right: 3vh;
    }

    .RecipeCard_full .closePopup {
        left: 1.2vw;
    }

    .RecipeCard_full .closePopup:active {
        box-shadow: none;
        left: 1vw;
    }
}