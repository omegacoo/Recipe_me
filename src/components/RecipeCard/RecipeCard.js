import React from 'react';

import StoreContext from '../../StoreContext';

import './RecipeCard.css';

export default class RecipeCard extends React.Component {
    static contextType = StoreContext;

    state = {
        currentRecipe: '',
        cardOpen: false
    };

    toggleCardOpen = e => {
        this.setState({
            currentRecipe: e.target.id,
            cardOpen: !this.state.cardOpen
        });
    };

    renderIngredients = () => {
        let ingredients = this.props.recipe.ingredients;

        if(this.context.loggedIn){
            ingredients = this.props.recipe.ingredients;
            ingredients = ingredients.map(i => i = {...i, ...this.context.ingredients.find(I => I.id === i.ingredient_id)});
            
            return ingredients.map(i => 
                <li key={i.ingredient_id} id={i.ingredient_id}>{i.quantity} - {i.unit} - {i.title}</li>    
            )
        } else {
            ingredients = ingredients.map(i => i = this.context.guestIngredients.find(gI => gI.id === i));
            
            return ingredients.map(i => 
                <li key={i.id} id={i.id}>{i.title}</li>
            );
        };
    };

    renderInstructions = () => {
        const instructions = this.props.recipe.instructions;
        let newInstructions = [];
        let count = 2;
        newInstructions.push('<mark>');
        for(let i = 0; i < instructions.length; i++){
            if(instructions[i] === count.toString() && instructions[i + 1] === '.'){
                newInstructions.push('\n\n<mark>' + instructions[i]);
                count++;
            } else if(instructions[i] === ':'){
                newInstructions.push(':</mark>');
            } else {
                newInstructions.push(instructions[i])
            };
        };
        return <p className='Recipe_instructions' dangerouslySetInnerHTML={{ __html: newInstructions.join('')}} />
    };

    renderView = () => {
        if(!this.state.cardOpen){
            return( 
                <li 
                    className='RecipeCard' 
                    id={this.props.recipe.id}
                    onClick={this.toggleCardOpen}
                >
                    <section className='RecipeCard_title' id={this.props.recipe.id}>{this.props.recipe.title}</section>
                    <section className='RecipeCard_description' id={this.props.recipe.id}>{this.props.recipe.description}</section>
                </li>
            )
        } else {
            return(
                <div className='RecipeCard_full'>
                    <i className='closePopup far fa-times-circle' onClick={this.toggleCardOpen} />
                    <div className='title_container'><h2 className='RecipeCard_full_title'>{this.props.recipe.title}</h2></div>
                    <section className='Recipe_times'>
                        <span className='prep_time'>prep time: {this.props.recipe.prep_time}</span>
                        <span className='cook_time'>cook time: {this.props.recipe.cook_time}</span>
                    </section>
                    <label className='Recipe_ingredients_label' htmlFor='Recipe_ingredients'>Ingredients</label>
                    <ul className='Recipe_ingredients'>
                        {this.renderIngredients()}
                    </ul>
                    <label className='Recipe_instructions_label' htmlFor='Recipe_instructions'>Instructions</label>
                    {this.renderInstructions()}
                </div>
            )
        };
    };

    render(){
        return <>{this.renderView()}</>
    };
};