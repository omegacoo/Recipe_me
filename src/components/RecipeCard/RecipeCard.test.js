import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import { BrowserRouter } from 'react-router-dom';
import RecipeCard from './RecipeCard';

describe('RecipeCard component', () => {
    let testRecipe = {
        id: '3',
        title: 'Bruschetta with tomato and basil',
        description: `Chopped fresh tomatoes with garlic, basil, olive oil, and vinegar, served on toasted slices of french bread.`,
        diet: null,
        instructions: `1. Blanch and peel the tomatoes: bring 2 quarts of water to a boil. As the water is heating make shallow cuts in a cross pattern at the tip ends of the tomatoes (this will make the tomatoes easier to peel). Once water is boiling, remove the pot from the heat. Put the tomatoes in the hot water and blanch for 1 minute. Remove with a slotted spoon and let sit until cool enough to handle. Then genlty peel off the tomato skins. Cut out the stem base with a paring knife. Cut the tomatoes into halves or quarters and squeeze out most of the juices and seeds. 2. Preheat oven: to 450 degrees with a rack in the top slot of the oven. 3. Chop tomatoes, toss them with garlics, olive oil, vinegar, basil, salt, and pepper: finely chop the tomatoes and place them in a medium bowl. Mix in the minced garlic, 1 Tbsp olive oil, and the balsamic vinegar. Stir in the thingly sliced basil and add salt and freshly ground black pepper, adding more to taste. Note, tomatoes love salt; you may need to add more than you expect. 4. Toast the baguette slices: use a bread knife to slice the baguette on the diagonal, making half-inch thick slices. Brush one side of each slice with olive oil (a pastry brush helps here) and place olive oil-side down on a baking sheet or roasting pan. The baguette slices will toast best in the top rack of your oven, so you may need to work in batches to toast them all. When the over has readh 450 degrees place the slices in the oven on the top rack and toast for 5 to 6 minutes until lightly browned around the edges. 5. Serve toasted bread with tomato mixture: arrange the toasted bread on a platter, olive oil side facing up (the olive oil will help create a temprorary barrier keeping the bread from getting soggy from the chopped tomatoes). Either serve the toasts plain with a bowl of the tomato brushetta mixture on the side for people to top their own, or sue a spoon to gently top each toasted bread slice with some of the tomato mixture. If you top each slice individually, do it right before serving.`,
        prep_time: '15 minutes',
        cook_time: '120 minutes',
        ingredients: [14, 15, 16, 17, 18, 9, 19, 20]
    };

    it('renders without crashing', () => {
        const div = document.createElement('div');

        ReactDOM.render(
            <BrowserRouter>
                <RecipeCard recipe={ testRecipe } />
            </BrowserRouter>, 
            div
        );
        ReactDOM.unmountComponentAtNode(div);
    });

    it('renders UI correctly', () => {
        expect(
        renderer.create(
            <BrowserRouter>
                <RecipeCard recipe={ testRecipe } />
            </BrowserRouter>).toJSON()
        ).toMatchSnapshot();
    });
});