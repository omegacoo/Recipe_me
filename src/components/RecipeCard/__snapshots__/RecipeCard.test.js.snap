// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RecipeCard component renders UI correctly 1`] = `
<li
  className="RecipeCard"
  id="3"
  onClick={[Function]}
>
  <section
    className="RecipeCard_title"
    id="3"
  >
    Bruschetta with tomato and basil
  </section>
  <section
    className="RecipeCard_description"
    id="3"
  >
    Chopped fresh tomatoes with garlic, basil, olive oil, and vinegar, served on toasted slices of french bread.
  </section>
</li>
`;
